includes/class-magic-link.php

```diff
@@ -0,0 +1,801 @@
+
+		/**
+		 * Filters whether the user can use magic links.
+		 *
+		 * @param bool     $can_magic_link Whether the user can use magic links.
+		 * @param \WP_User $user           User object.
+		 */
+		return \apply_filters( 'newspack_can_magic_link', $can_magic_link, $user );
+	}
+
+	/**
+	 * Get magic link token expiration period.
+	 *
+	 * @return int Expiration in seconds.
+	 */
+	private static function get_token_expiration_period() {
+		/**
+		 * Filters the duration of the magic link token expiration period.
+		 *
+		 * @param int    $length Duration of the expiration period in seconds.
+		 */
+		return \apply_filters( 'newspack_magic_link_token_expiration', 30 * MINUTE_IN_SECONDS );
+	}
+
+	/**
+	 * Clear client secret cookie.
+	 */
+	public static function clear_client_secret_cookie() {
+		/** This filter is documented in wp-includes/pluggable.php */
+		if ( ! apply_filters( 'send_auth_cookies', true ) ) {
+			return;
+		}
+
+		if ( ! isset( $_COOKIE[ self::COOKIE ] ) ) {
+			return;
+		}
+
+		// phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.cookies_setcookie
+		setcookie( self::COOKIE, ' ', time() - YEAR_IN_SECONDS, COOKIEPATH, COOKIE_DOMAIN );
+	}
+
+	/**
+	 * Get locally stored secret to be used as part of the client hash.
+	 *
+	 * @param bool $reset Whether to generate a new secret.
+	 */
+	private static function get_client_secret( $reset = false ) {
+		$secret = self::$session_secret;
+
+		/** Fetch cookie if available. */
+		if ( empty( $secret ) && isset( $_COOKIE[ self::COOKIE ] ) ) {
+			// phpcs:ignore WordPressVIPMinimum.Variables.RestrictedVariables.cache_constraints___COOKIE
+			$secret = \sanitize_text_field( \wp_unslash( $_COOKIE[ self::COOKIE ] ) );
+		}
+
+		/** Regenerate if empty or resetting. */
+		if ( empty( $secret ) || true === $reset ) {
+			$secret = \wp_generate_password( 43, false, false );
+		}
+
+		self::$session_secret = $secret;
+
+		/** This filter is documented in wp-includes/pluggable.php */
+		if ( \apply_filters( 'send_auth_cookies', true ) ) {
+			/** Add an extra 5 minutes to the client secret cookie expiration. */
+			$expiration = time() + self::get_token_expiration_period() + ( 5 * MINUTE_IN_SECONDS );
+
+			// phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.cookies_setcookie
+			setcookie( self::COOKIE, $secret, $expiration, COOKIEPATH, COOKIE_DOMAIN, true );
+		}
+
+		return $secret;
+	}
+
+	/**
+	 * Get client hash for the current session.
+	 *
+	 * @param \WP_User $user         User the client hash is being generated for.
+	 * @param bool     $reset_secret Whether to reset the stored client secret.
+	 *
+	 * @return string|null Client hash or null if unable to generate one.
+	 */
+	private static function get_client_hash( $user, $reset_secret = false ) {
+		/** Don't return client hash from CLI command */
+		if ( defined( 'WP_CLI' ) ) {
+			return null;
+		}
+
+		/** Don't return client hash if it's not self-served. */
```

Comment:
Double negation always add a few nanoseconds to processing time in a human brain, let's keep it simple:

```suggestion
		/** Return client hash only if it's self-served. */
```